### PR TITLE
Modal dialogs async

### PR DESCRIPTION
**Pull request by vladest**
case: https://highfidelity.fogbugz.com/f/cases/6821/Modal-dialog-s-blocks-the-thread-that-creates-them
The PR is for preliminary understanding of things gets right

Test plan:
- Goto Toolbar mode
- Try to do activities with MessageBox dialogs gets opened. Mostly within Marketplace
- Goto Tablet mode
- make sure everything still works
- Test file open dialogs: in Running Scripts try to load scripts from disk
- Test input fields: in Running Scripts try to load scripts from URL

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2408/HighFidelity-Beta-PR11159-b780d4bce3120a39838963163c869724bcf4ad2c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2408/HighFidelity-Beta-PR11159-b780d4bce3120a39838963163c869724bcf4ad2c.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2494/HighFidelity-Beta-PR11159-9aace9e67d4f9bacc35c8b30de4b554398b0ae41.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2510/HighFidelity-Beta-PR11159-273261ee5705314b7b6f7ae31d201225abf7f99e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2510/HighFidelity-Beta-PR11159-273261ee5705314b7b6f7ae31d201225abf7f99e.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2552/HighFidelity-Beta-PR11159-15f8bc0141f993ffec8f651bc74f9956938c1eb8.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2552/HighFidelity-Beta-PR11159-15f8bc0141f993ffec8f651bc74f9956938c1eb8.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2836/HighFidelity-Beta-PR11159-38295ab73f0d6b8044a9633d1746e74ffedc2217.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2836/HighFidelity-Beta-PR11159-38295ab73f0d6b8044a9633d1746e74ffedc2217.exe

**Comment by Nex-Pro** [comment]
Testing this as requested.

**Comment by Nex-Pro** [comment]
Testing setup:

OS: Win10
CPU: i7 @ 4.0ghz
Ram: 16gb
GPU: Nvidia GTX1070
HMD: Vive

### Possibly Fail read results.

Toolbar mode dialogs.

Menu/File/login| **SUCCESS**
Menu/Edit/Running Scripts| **issues seen not due to this PR** Will check fogbugz.
Menu/Edit/Running Scripts/From URL| **SUCCESS**
Menu/Edit/Running Scripts/From Disk| **SUCCESS**
Menu/Edit/Open and run script file| **SUCCESS**
Menu/Edit/Open and run script from URL| **SUCCESS**
Menu/Edit/Asset Browser/choose model| **SUCCESS**
Menu/Edit/Package Model ...| **SUCCESS**
Menu/Edit/Import Entities| **SUCCESS**
Menu/Edit/Import Entities From URL| **FAIL** Nothing imported

Menu/Avatar/Attachments| **SUCCESS**
Menu/Avatar/Bookmark Avatar| **SUCCESS**
Menu/Avatar/Delete Avatar Bookmark| **SUCCESS**

Menu/Navigate/Bookmark Location| **SUCCESS**
Menu/Navigate/Bookmark Location (Overwrite)| **SUCCESS**
Menu/Navigate/Delete Bookmark| **SUCCESS**

Snapshot/Choose folder| **FAIL** Does not choose the folder correctly

Goto/Type address| **SUCCESS**

Marketplace/Search| **SUCCESS**
Marketplace/Accept terms modal ( photo real avatars )| **SUCCESS**
Marketplace/Login to Clara IO| **SUCCESS**
Marketplace/Clara IO/ Download Model/Cancel| **SUCCESS**

Tablet mode dialogs.

Menu/File/login| **SUCCESS**
Menu/Edit/Running Scripts| **issues seen not due to this PR** Will check fogbugz.
Menu/Edit/Running Scripts/From URL| **SUCCESS**
Menu/Edit/Running Scripts/From Disk| **SUCCESS**
Menu/Edit/Open and run script file| **SUCCESS**
Menu/Edit/Open and run script from URL| **SUCCESS**
Menu/Edit/Asset Browser/choose model| **SUCCESS**
Menu/Edit/Package Model ...| **SUCCESS**
Menu/Edit/Import Entities| **SUCCESS**
Menu/Edit/Import Entities From URL| **FAIL** Nothing imported

Menu/Avatar/Attachments| **SUCCESS**
Menu/Avatar/Bookmark Avatar| **SUCCESS**
Menu/Avatar/Delete Avatar Bookmark| **SUCCESS**

Menu/Navigate/Bookmark Location| **SUCCESS**
Menu/Navigate/Bookmark Location (Overwrite)| **SUCCESS**
Menu/Navigate/Delete Bookmark| **SUCCESS**

Snapshot/Choose folder| **FAIL** Does not choose the folder correctly

Goto/Type address| **SUCCESS**

Marketplace/Search| **SUCCESS**
Marketplace/Accept terms modal ( photo real avatars )| **SUCCESS**
Marketplace/Login to Clara IO| **SUCCESS**
Marketplace/Clara IO/ Download Model/Cancel| **SUCCESS**

**Comment by anshuman64** [comment]
@Nex-Pro Calling this a QA Pass. Those two issues exist in current master.

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -6191,50 +6191,68 @@ bool Application::askToSetAvatarUrl(const QString& url) {
...
+    bool agreeToLicense = true; // assume true
+    //create set avatar callback
+    auto setAvatar = [=] (QString url, QString modelName) {
+        auto offscreenUi = DependencyManager::get<OffscreenUi>();
 
-        agreeToLicence = QMessageBox::Yes == OffscreenUi::question("Avatar Usage License",
-            modelLicense + "\nDo you agree to these terms?",
-            QMessageBox::Yes | QMessageBox::No, QMessageBox::Yes);
-    }
+        QObject::connect(offscreenUi.data(), &OffscreenUi::response, this, [=] (QMessageBox::StandardButton answer) {
```
Comment (jherico) [inline interface/src/Application.cpp]: The approach of connecting to a `OffscreenUi::response` isn't going to work.  If multiple scripts create dialogs in this fashion there's no way for them determine whether the `response` signal they get is in response to their dialog or to another that was created by someone else.

Instead, a more reasonable approach that should let you keep most of this code would be to have the async functions return an object which contains a response signal and a `show` method.

So...

```
// Creates the dialog but doesn't yet display it
AsyncDialog* query = OffscreenUi::asyncQuestion(...);
// Connect to the signal sent on dialog completion
connect(query, &AsyncDialog::response, ...)
// Show the dialog
query->show();
```

Comment (vladest) [reply interface/src/Application.cpp]: so, basically, each async dialog query should create an object instance, and all of them should be manged somehow? Within OffscreenUi?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3380/HighFidelity-Beta-PR11159-1cde504c741d146ec9751aac9a6883ce6ba1c13f.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3380/HighFidelity-Beta-PR11159-1cde504c741d146ec9751aac9a6883ce6ba1c13f.exe

**Comment by anshuman64** [comment]
@Nex-Pro can we get another pass on this?

**Comment by anshuman64** [comment]
@jherico This is ready for code review again

**Comment by Nex-Pro** [comment]
retesting as per request.

**Comment by Nex-Pro** [comment]
## FAIL

### Testing setup:

#### OS: Win10
#### CPU: i7 @ 4.0ghz
#### Ram: 16gb
#### GPU: Nvidia GTX1070
#### HMD: Vive
#### AUDIO: USB headset


#### Testing conditions:
Complete uninstall of previous build and removal of folders in local and roaming.

### ========================================================================

### Toolbar mode dialogs.

Menu/File/login: **SUCCESS**
Menu/Edit/Running Scripts: **SUCCESS**
Menu/Edit/Running Scripts/From URL: **SUCCESS**
Menu/Edit/Running Scripts/From Disk:: **FAIL** No messages in log. Script does not load.
Menu/Edit/Open and run script file: **FAIL** No messages in log. Script does not load.
Menu/Edit/Open and run script from URL: **SUCCESS**
Menu/Edit/Asset Browser/choose model: **SUCCESS**
Menu/Edit/Package Model ...: **SUCCESS**
Menu/Edit/Entity list: **FAIL** Nothing happens.
Menu/Edit/Import Entities: **FAIL** Get error message.
Menu/Edit/Import Entities From URL: **FAIL** Nothing imported

Log.. 

[09/19 13:35:13] [CRITICAL] [hifi.scriptengine] [UncaughtException signalHandlerException] Error: Function.prototype.disconnect: failed to disconnect from WindowScriptingInterface::openFileChanged(QString) in file:///E:/Program Files/High Fidelity - PR11159/scripts/system/edit.js:1466
[09/19 13:35:13] [CRITICAL] [hifi.scriptengine] [Backtrace]
[09/19 13:35:13] [CRITICAL] [hifi.scriptengine]     onFileOpenChanged(filename = 'file:///C:/Users/NexPro/Documents/hifi/Dodecasculpture.json') at file:///E:/Program Files/High Fidelity - PR11159/scripts/system/edit.js:1466
[09/19 13:35:13] [CRITICAL] [hifi.scriptengine]     <global>() at -1

Menu/Avatar/Attachments: **SUCCESS** But see notes.
Menu/Avatar/Bookmark Avatar: **SUCCESS**
Menu/Avatar/Delete Avatar Bookmark: **SUCCESS**

Menu/Navigate/Bookmark Location: **SUCCESS**
Menu/Navigate/Bookmark Location (Overwrite): **FAIL** Invokes crash every time ( is fine in HMD tablet ) no splats generated.
See notes for other issues re: bookmarks

Menu/Navigate/Delete Bookmark: **SUCCESS**

Snapshot/Choose folder: **FAIL** Does not choose the folder correctly but can change afterwards.

Goto/Type address: **SUCCESS**

Marketplace/Search: **SUCCESS**
Marketplace/Accept terms modal ( photo real avatars ): **SUCCESS**
Marketplace/Login to Clara IO: **FAIL** Login can be blocked (see image)
hifi-snap-by-piper-peppercorn-on-2017-09-19_14-11-08


Marketplace/Clara IO/ Download Model/Cancel: **FAIL** Downloads model and adds single copy to world.

### Tablet mode dialogs.

Menu/File/login: **SUCCESS**
Menu/Edit/Running Scripts

 After attempting to choose a script by typing "debug" into the search field I was unable to interact with the tablet other than open and close. ( reloaded all scripts via menu in desktop mode, )

Menu/Edit/Running Scripts/From URL: **SUCCESS**
Menu/Edit/Running Scripts/From Disk: **FAIL**
Menu/Edit/Open and run script file: **FAIL** Could not open any scripts
Menu/Edit/Open and run script from URL: **SUCCESS**
Menu/Edit/Asset Browser/choose model: **SUCCESS**

Menu/Edit/Package Model ...
: **FAIL 1** Dialog opened on second monitor. About the size of a postage stamp on a 4k. 
: **FAIL 2** Crash on closing dialog.
: **FAIL 3** Can keep clicking and getting another package dialog for each click.
Menu/Edit/Entity list: **FAIL** Nothing happens.
Menu/Edit/Import Entities: **FAIL** Nothing imported plus got notification about an error. see image. Also could not scroll menu after this ( had to reload scripts from app menu )
Menu/Edit/Import Entities From URL: **FAIL** Nothing imported plus interaction with tablet failed ( had to reload scripts from app menu again )

Menu/Avatar/Attachments: **SUCCESS** Note: took several attempts to delete attachment.
Menu/Avatar/Bookmark Avatar: **SUCCESS**
Menu/Avatar/Delete Avatar Bookmark: **SUCCESS** 

Menu/Navigate/Bookmark Location: **SUCCESS** 
Menu/Navigate/Bookmark Location (Overwrite): **SUCCESS** But caused tablet interactions to fail.
Menu/Navigate/Delete Bookmark: **SUCCESS** 

Snapshot/Choose folder: **FAIL** Does not choose the folder correctly. Known issue.

Goto/Type address: **SUCCESS** Unable to insert a character between others on this dialog. Can do on others.

Marketplace/Search: **SUCCESS**
Marketplace/Accept terms modals ( photo real avatars ): **SUCCESS** 
Marketplace/Login to Clara IO: **SUCCESS**
Marketplace/Clara IO/ Download Model/Cancel: **FAIL** Not due to this PR but is placing the model in world twice for each asset downloaded from Clara.IO may not have noticed this first pass on this PR. Will check on master/fogbugz


Notes:

On changing to HMD I got midnight sky situation when I should have seen the regular localhost build platform.
Will check fogbugz for this. Cleared with ctrl+1 then ctrl + 4.

User name not displaying correctly on tablet. See image.
hifi-snap-by-piper-peppercorn-on-2017-09-19_11-31-24


Edit Menu layout breaks in HMD. See image. Seems to be caused by reloading scripts.
hifi-snap-by-piper-peppercorn-on-2017-09-19_14-28-58

We need to stop returning to the main menu between operations. This makes managing bookmarks etc. tedious.

Had to resort to task manager to quit interface on one occasion. ( ***probably a windows issue as alt+f4 also did not work*** )

Navigate -> copy path to clipboard   seems incorrectly named. It copies the local position and orientation. Thus: /1.17393,-200.288,-7.93192/0,0.923287,0,0.384111

Unable to drag->drop .JSON files to asset server
.JSON files added via 'choose' in asset browser cannot be added to world.

copy URL for selected asset in asset browser does not work. 

Can't paste anything into the 'choose' dialog of the asset browser.

Text pasted into the avatar attachments dialog is not validated in any way.

Invalid text in attachment dialog URL is retained.

Again the setting for snap turn gets forgotten after a crash. Advanced movement is not forgotten.

Using bookmarks to jump to a remote works. However the avatar becomes disconnected when subsequently using a local bookmark.

**Comment by vladest** [comment]
File loading issues must be fixed now

**Comment by Nex-Pro** [comment]
re-testing this shortly

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3447/HighFidelity-Beta-PR11159-50ee41e0955982bacd3141ee98be81ef76a3a1dd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3447/HighFidelity-Beta-PR11159-50ee41e0955982bacd3141ee98be81ef76a3a1dd.exe

**Comment by Nex-Pro** [comment]
Testing this with build ending in 3a1dd  ( clean install )

**Comment by Nex-Pro** [comment]
### FAIL on a couple of small points.

Menu/File/login- **SUCCESS**
Menu/Edit/Running Scripts:- **SUCCESS**
Menu/Edit/Running Scripts/From URL:- **SUCCESS**
Menu/Edit/Running Scripts/From Disk:- **SUCCESS**
Menu/Edit/Open and run script file:- **SUCCESS**
Menu/Edit/Open and run script from URL:- **SUCCESS**
Menu/Edit/Asset Browser/choose model:- **SUCCESS** 
Menu/Edit/Package Model:- **SUCCESS** 
Menu/Edit/Entity list:- **FAIL Nothing happens.**
Menu/Edit/Import Entities:- **SUCCESS**

Menu/Edit/Import Entities From URL:- **FAIL**  ***Nothing imported***

log output:

[09/19 23:27:43] [CRITICAL] [hifi.scriptengine] [UncaughtException signalHandlerException] Error: Function.prototype.disconnect: failed to disconnect from WindowScriptingInterface::openFileChanged(QString) in file:///E:/Program Files/High Fidelity - PR11159/scripts/system/edit.js:1466
[09/19 23:27:43] [CRITICAL] [hifi.scriptengine] [Backtrace]
[09/19 23:27:43] [CRITICAL] [hifi.scriptengine]     onFileOpenChanged(filename = 'http://mpassets.highfidelity.com/6af97d4e-1338-4b4f-97bd-4d80449d5542-v1/circusHammer.SVO.json') at file:///E:/Program Files/High Fidelity - PR11159/scripts/system/edit.js:1466
[09/19 23:27:43] [CRITICAL] [hifi.scriptengine]     <global>() at -1


Menu/Avatar/Attachments:- **SUCCESS**
Menu/Avatar/Bookmark Avatar:- **SUCCESS**
Menu/Avatar/Delete Avatar Bookmark:- **SUCCESS**

Menu/Navigate/Bookmark Location:- **SUCCESS**

Menu/Navigate/Bookmark Location (Overwrite):- **FAIL**  ***Still Invokes crash every time ( is fine in HMD tablet ) no splats generated.***

Menu/Navigate/Delete Bookmark:- **SUCCESS**
***See previous notes for other issues re: bookmarks***

Snapshot/Choose folder:- **Worked this time (no subfolders in the selected folder)**

Goto/Type address:- **SUCCESS**

Marketplace/Search:- **SUCCESS**
Marketplace/Accept terms modal ( photo real avatars ):- **SUCCESS**
Marketplace/Login to Clara IO:- **FAIL** ***Login can be blocked (see previously posted image)***


Marketplace/Clara IO/ Download Model/Cancel:- **SUCCESS** Downloads model and adds single copy to world. (may have erroneously noted this as a fail in previous results)


## Tablet mode dialogs.

Menu/File/login:- **SUCCESS**


### Menu/Edit/Running Scripts

After attempting to choose a script by typing "debug" into the search field I was unable to interact with the tablet other than open and close. ( reloaded all scripts via menu in desktop mode, )

Menu/Edit/Running Scripts/From URL:- **SUCCESS**
Menu/Edit/Running Scripts/From Disk:- **SUCCESS**
Menu/Edit/Open and run script file:- **SUCCESS**
Menu/Edit/Open and run script from URL:- **SUCCESS**
Menu/Edit/Asset Browser/choose model:- **SUCCESS**

Menu/Edit/Package Model ...
: **FAIL** Can keep clicking and getting another package dialog for each click.

Menu/Edit/Entity list:- As previous **FAIL** ***Nothing happens.***
Menu/Edit/Import Entities:- **SUCCESS**
Menu/Edit/Import Entities From URL:- **FAIL** ***Nothing imported***

Menu/Avatar/Attachments:- **SUCCESS**
Menu/Avatar/Bookmark Avatar:- **SUCCESS**
Menu/Avatar/Delete Avatar Bookmark:- **SUCCESS**

Menu/Navigate/Bookmark Location:- **SUCCESS**
Menu/Navigate/Bookmark Location (Overwrite):- **SUCCESS**
Menu/Navigate/Delete Bookmark:- **SUCCESS**

Snapshot/Choose folder:- **SUCCESS** Apart from known issue.

Goto/Type address:- **SUCCESS** but still unable to insert a character between others on this dialog. Can do on other dialogs.

Marketplace/Search:- **SUCCESS**
Marketplace/Accept terms modals ( photo real avatars ):- **SUCCESS**
Marketplace/Login to Clara IO:- **SUCCESS**
Marketplace/Clara IO/ Download Model/Cancel:- **FAIL** ***Not due to this PR but is placing the model in world twice for each asset downloaded from Clara.***

Got a situation a few times where I was left with a private tablet in front of me. Hitting the tablet button on the controller removed it and showed the correct one. This has been reported by others.

Had to restart scripts several times ( way too often )
Had several crashes no splats.
Had to restart interface severl times.
**GRIPES:** Probably not due to this PR.
New position that the tablet appears at is really ***(insert stream of expletives here)*** annoying( Who's idea was it to move it from near the hand ?)

**Comment by vladest** [comment]
Fixed crash on second bookmark saving and Import Entities from URL
Entities list doesnt works on master as well. need separate fix
Clara login is part of web code which is not related to the PR

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3475/HighFidelity-Beta-PR11159-6e3cfe81ab5288cbef675e8b9f492e66900c1b56.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3475/HighFidelity-Beta-PR11159-6e3cfe81ab5288cbef675e8b9f492e66900c1b56.exe

**Comment by anshuman64** [comment]
@vladest Great work thanks. Created tickets for things broken in master. 

@jherico Can we get a CR please?

**Pull request merged**